### PR TITLE
[TACHYON-570] Adding documentation and tests to what was CommonUtils.

### PR DESCRIPTION
**Pull request by yuzhu**
https://tachyon.atlassian.net/browse/TACHYON-570

Common utility functions are split up into several utils files.
So are their tests.

### CODE REVIEW

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java [hunk 1]
```diff
@@ -16,13 +16,41 @@
...
+      Assert.assertArrayEquals(new String[] {"", "bar"}, PathUtils.getPathComponents("/bar"));
+      Assert.assertArrayEquals(new String[] {"", "foo", "bar"},
+          PathUtils.getPathComponents("/foo/bar"));
+      Assert.assertArrayEquals(new String[] {"", "foo", "bar"},
+          PathUtils.getPathComponents("/foo/bar/"));
+      Assert.assertArrayEquals(new String[] {"", "bar"},
+          PathUtils.getPathComponents("/foo/../bar"));
+      Assert.assertArrayEquals(new String[] {"", "foo", "bar", "a", "b", "c"},
+          PathUtils.getPathComponents("/foo//bar/a/b/c"));
+
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: extra empty line

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java [hunk 2]
```diff
@@ -16,13 +16,41 @@
...
 import tachyon.Constants;
+import tachyon.thrift.InvalidPathException;
 
 public class PathUtilsTest {
+  @Rule
+  public final ExpectedException mException = ExpectedException.none();
+
+  @Test
+  public void getPathComponentsTest() throws InvalidPathException {
+    try {
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: NIT: better to split this testcase into two testcases: one is the test with successful results (i.e., Asserts within the try-catch block), and one test is expecting the exception.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  @Rule
+  public final ExpectedException mException = ExpectedException.none();
+
+  /**
+   * Test for both changeLocalFilePermission and changeLocalFileToFullPermission
+   *
+   * @param filePath that will change permission
+   * @throws IOException
+   */
+
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: extra line.

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 4]
```diff
@@ -42,10 +42,26 @@ public static void cleanDirectBuffer(ByteBuffer buffer) {
...
+  /**
+   * Clone a bytebuffer.
+   * <p>
+   * The new bytebuffer will have the same content, but the cloning only applies to array-based
+   * bytebuffers
+   *
+   * @param buf The ByteBuffer to clone
+   * @return The new ByteBuffer
+   */
+
```
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: extra empty line

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 5]
```diff
@@ -42,10 +42,26 @@ public static void cleanDirectBuffer(ByteBuffer buffer) {
...
   public static ByteBuffer cloneByteBuffer(ByteBuffer buf) {
     ByteBuffer ret = ByteBuffer.allocate(buf.limit() - buf.position());
-    ret.put(buf.array(), buf.position(), buf.limit() - buf.position());
-    ret.flip();
+    if (buf.hasArray()) {
+      ret.put(buf.array(), buf.position(), buf.limit() - buf.position());
+      ret.flip();
```
Comment (apc999) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: NIT: both if cases has `ret.flip()`, maybe consolidate these two after if-else ?

**Comment by apc999** [comment]
Thanks for improving our unit-tests! A few comments left

**Comment by yuzhu** [comment]
addressed your comments, @apc999

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java [hunk 6]
```diff
@@ -16,13 +16,43 @@
...
 import tachyon.Constants;
+import tachyon.thrift.InvalidPathException;
 
 public class PathUtilsTest {
+  @Rule
+  public final ExpectedException mException = ExpectedException.none();
+
+  @Test
+  public void getPathComponentsTestSuccess() {
+    try {
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: Do we still need `try`-`catch` here?
It seems that the test fails anyway when `InvalidPathException` is thrown.

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java [hunk 7]
```diff
@@ -16,13 +16,43 @@
...
+          PathUtils.getPathComponents("/foo/../bar"));
+      Assert.assertArrayEquals(new String[] {"", "foo", "bar", "a", "b", "c"},
+          PathUtils.getPathComponents("/foo//bar/a/b/c"));
+    } catch (InvalidPathException e) {
+      Assert.fail();
+    }
+  }
+
+  @Test
+  public void getPathComponentsTestException() throws InvalidPathException {
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: Tachyon convention is to name testcase suffixed with `Test` :)
Same comment applies to above testcase.

**Comment by apc999** [comment]
Two more minor comments here left. Thanks for the update!

File: common/src/main/java/tachyon/util/io/BufferUtils.java [hunk 8]
```diff
@@ -42,9 +42,23 @@ public static void cleanDirectBuffer(ByteBuffer buffer) {
     }
   }
 
+  /**
+   * Clone a bytebuffer.
+   * <p>
+   * The new bytebuffer will have the same content, but the cloning only applies to array-based
+   * bytebuffers
```
Comment (calvinjia) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: What do you mean by `the cloning only applies to array-based bytebuffers`?

File: common/src/test/java/tachyon/util/io/ByteUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    ByteBuffer bufClone = null;
+    for (byte i = 0; i < bufferSize; i ++) {
+      buf.put(i);
+      bufDirect.put((byte) (i + 1));
+    }
+    bufClone = BufferUtils.cloneByteBuffer(buf);
+    Assert.assertTrue(bufClone.equals(buf));
+    bufClone = BufferUtils.cloneByteBuffer(bufDirect);
+    Assert.assertTrue(bufClone.equals(bufDirect));
+  }
```
Comment (calvinjia) [inline common/src/test/java/tachyon/util/io/ByteUtilsTest.java]: Could you split this into two different tests?

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    // Change permission on directories
+    FileUtils.changeLocalFilePermission(mTestFolder.getRoot().getAbsolutePath(), "444");
+    Assert.assertFalse(tempFile.delete());
+    FileUtils.changeLocalFilePermission(mTestFolder.getRoot().getAbsolutePath(), "744");
+    Assert.assertTrue(tempFile.delete());
+
+    // tempFile no longer exist, so changing permission on it should return exception
+    mException.expect(IOException.class);
+    FileUtils.changeLocalFilePermission(tempFile.getAbsolutePath(), "777");
+  }
```
Comment (calvinjia) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Could you split this into two different tests?

Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: +1 , ideally, each test case should be a simple one to cover one case --- though sometimes we piggyback test cases to save setup overhead. But it seems in this particular test case, we are pretty fine.

File: common/src/main/java/tachyon/util/io/BufferUtils.java (hunk 8 above)
Comment (yuzhu) [inline common/src/main/java/tachyon/util/io/BufferUtils.java]: what i meant is the cloned buffer may be a different type.. will change it to say that.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 10 above)
Comment (yuzhu) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: The comment is incorrect. This is only testing changeLocalFilePermission.

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java (hunk 7 above)
Comment (yuzhu) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: ok

File: common/src/test/java/tachyon/util/io/PathUtilsTest.java (hunk 6 above)
Comment (yuzhu) [inline common/src/test/java/tachyon/util/io/PathUtilsTest.java]: agree

File: common/src/test/java/tachyon/util/io/ByteUtilsTest.java (hunk 9 above)
Comment (yuzhu) [inline common/src/test/java/tachyon/util/io/ByteUtilsTest.java]: yup

Comment (yuzhu) [inline common/src/test/java/tachyon/util/io/ByteUtilsTest.java]: yes

File: common/src/main/java/tachyon/util/FormatUtils.java [hunk 11]
```diff
@@ -58,6 +58,12 @@ public static void printTimeTakenNs(long startTimeNs, Logger logger, String mess
     logger.info(message + " took " + (System.nanoTime() - startTimeNs) + " ns.");
   }
 
+  /**
+    * Return a human-readable version of bytes 10GB 2048KB etc.
```
Comment (apc999) [inline common/src/main/java/tachyon/util/FormatUtils.java]: line 62-66 of javadoc has wrong indent.

Comment (yuzhu) [inline common/src/main/java/tachyon/util/FormatUtils.java]: fixed

**Comment by apc999** [comment]
+1 LGTM

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 10 above)
Comment (calvinjia) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: I thought there are two different tests, one for a file changing permission and one for a directory?

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    // expect a file permission error when we open it for writing
+    mException.expect(IOException.class);
+    FileWriter fw = new FileWriter(tempFile);
+    if (fw != null) {
+      fw.close();
+    }
+    // tempFile no longer exist, so changing permission on it should return exception
+    mException.expect(IOException.class);
+    FileUtils.changeLocalFilePermission(tempFile.getAbsolutePath(), "777");
+    Assert.assertTrue(tempFile.delete());
```
Comment (calvinjia) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Do you think it makes sense to make this also another test (the change nonexistent file test), sorry for not bringing this up earlier.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+public class FileUtilsTest {
+  @Rule
+  public TemporaryFolder mTestFolder = new TemporaryFolder();
+  @Rule
```
Comment (calvinjia) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Could you add a space between the `@Rule`s? I think that's the style we have currently.

**Comment by calvinjia** [comment]
@yuzhu Looks good overall, just two minor comments.

**Comment by calvinjia** [comment]
+1 LGTM

**Pull request merged**